rfcs/0004-range-queries.md

```diff
@@ -0,0 +1,219 @@
+
+Range queries must access mutable table state, such as the current WAL. It is 
+therefore possible for the query results to include some records which 
+were written after the  query began scanning its range of records. For example,
+a query may include  only a subset of records that were inserted in the same 
+`WriteBatch`. In  other words, range queries will not have snapshot isolation 
+initially. 
+
+[RFC-????](https://github.com/slatedb/slatedb/pull/260) proposes transactional
+semantics for SlateDB. This relies on a monotonic sequence number which is tagged
+on each key update. Range queries could leverage this sequence number to improve
+isolation by filtering any updates with sequence numbers larger than the latest
+at the time the query was executed. We leave this as a gap for future work to address.
+
+### Checkpoint References
+
+We intend to rely on state checkpoints introduced in [RFC-????](https://github.com/slatedb/slatedb/pull/278) in order to prevent the cleanup
+of durable state referenced by the query. When a range scan is created, it will create
+an internal reference to the current checkpoint. This reference will prevent cleanup of
+the checkpoint state until the scan is complete. 
+
+Checkpoints created by `Readers` have an expiry time, which is only refreshed as long
+as it remains the current checkpoint. If a `DbIterator` instance remains active long 
+enough for its referenced checkpoint to expire, the `Reader` will proactively invalidate 
+it in  order to allow for safe cleanup. Invalidated iterators will return 
+`SlateDbError::InvalidatedIterator` on any subsequent call to `DbIterator::next`
+or `DbIterator::seek`.
+
+In general, we make the initial simplifying assumption that scans will not be frequent
+enough or last long enough for checkpoint accumulation to be a problem. In the future,
+we could either limit the number of checkpoint references that are available for
+scanning or we could limit the time that a `DbIterator` will remain valid.
+
+### Memory Usage
+
+In addition to the checkpoint, a range scan may hold references to records in unflushed 
+WALs or memtables. Rather than retaining references to these tables for the full duration
+of  the query, we can filter the set of records within the query's range.  
+As new records are returned, previous ones fall out of scope and may be cleaned up. This
+makes sense for bounded queries where the number of matching keys within the WAL
+or memtables may be low. For an unbounded query, we may prefer to hold a reference to the
+full table.
+
+Iterators will use prefetching to improve performance. Internally, this is 
+already implemented by `SstIterator`. Users can influence prefetching to improve
+performance using `ScanOptions::read_ahead_size`, which indicates the number of bytes
+to prefetch. Internally, this is used to set `SstIterator::blocks_to_fetch`. The 
+default `read_ahead_size` will be the size of a single block.
```

Comment:
Might be overkill, but do we want to expose `max_fetch_tasks` to `ScanOptions` as well?

Reply:
SGTM. I'm very sensitive to having too many configs, so I prefer excluding checkpoint expiry and max fetch tasks unless we _really_ need them. So for now, let's leave 'em out.